src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.Helpers.cs

```diff
@@ -0,0 +1,375 @@
+        private const string CollectionInterfaceTypeName = "System.Collections.ICollection";
+
+        private const string ReadOnlyListGenericInterfaceTypeName = "System.Collections.Generic.IReadOnlyList`1";
+
+        private const string ReadOnlyCollectionGenericInterfaceTypeName = "System.Collections.Generic.IReadOnlyCollection`1";
+
+        public const string HashtableTypeName = "System.Collections.Hashtable";
+        public const string SortedListTypeName = "System.Collections.SortedList";
+
+        public const string StackTypeName = "System.Collections.Stack";
+        public const string StackGenericTypeName = "System.Collections.Generic.Stack`1";
+
+        public const string QueueTypeName = "System.Collections.Queue";
+        public const string QueueGenericTypeName = "System.Collections.Generic.Queue`1";
+
+        public const string SetGenericInterfaceTypeName = "System.Collections.Generic.ISet`1";
+        public const string SortedSetGenericTypeName = "System.Collections.Generic.SortedSet`1";
+        public const string HashSetGenericTypeName = "System.Collections.Generic.HashSet`1";
+
+        public const string LinkedListGenericTypeName = "System.Collections.Generic.LinkedList`1";
+
+        public const string DictionaryInterfaceTypeName = "System.Collections.IDictionary";
+        public const string DictionaryGenericTypeName = "System.Collections.Generic.Dictionary`2";
+        public const string DictionaryGenericInterfaceTypeName = "System.Collections.Generic.IDictionary`2";
+        public const string ReadOnlyDictionaryGenericInterfaceTypeName = "System.Collections.Generic.IReadOnlyDictionary`2";
+        public const string SortedDictionaryGenericTypeName = "System.Collections.Generic.SortedDictionary`2";
+        public const string KeyValuePairGenericTypeName = "System.Collections.Generic.KeyValuePair`2";
+
+        public const string ArrayListTypeName = "System.Collections.ArrayList";
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary<,>),
+            typeof(ICollection<>),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_nonGenericInterfacesWithAddMethods = new Type[]
+        {
+            typeof(IDictionary),
+            typeof(IList),
+        };
+
+        // In the order we wish to detect a derived type.
+        private static readonly Type[] s_genericInterfacesWithoutAddMethods = new Type[]
+        {
+            typeof(IReadOnlyDictionary<,>),
+            typeof(IReadOnlyCollection<>),
+            typeof(IReadOnlyList<>),
+            typeof(IEnumerable<>),
+        };
+
+        public static Type GetImplementedCollectionType(Type queryType)
+        {
+            Debug.Assert(queryType != null);
+
+            if (queryType.IsAbstract ||
```

Comment:
Can we have a fast-path to exit if IEnumerable not implemented?

Reply:
Or use GetClassType() instead?

Reply:
I've reordered to have a fast-path exit for types that don't implement `IEnumerable` and `string` types. We still need to check for `string` because it implements `IEnumerable`.

Removed the check for `object`.

`GetClassType()` calls this method - using it would be an unwanted circular dependency.